csrc/dynamic_transform.cpp

```diff
@@ -385,6 +689,224 @@ void DynamicTransformConcretizer::concretize() {
   }
 }
 
+namespace {
+
+//! This function simply extracts the maybeExpandedExtent Vals from the
+//! noReductions(maybeRFactorDomain) of the provided tensorview.
+std::vector<Val*> tensor_shape(TensorView* tv) {
+  const auto& rfactor = TensorDomain::noReductions(tv->getMaybeRFactorDomain());
+  std::vector<Val*> shape;
+  shape.reserve(rfactor.size());
+  for (const auto id : rfactor) {
+    shape.push_back(id->getMaybeExpandedExtent());
+  }
+  return shape;
+}
+
+//! Return whether or not we should replace a reduction or Welford op that is
+//! given an empty input tensor. Returns false if any non-reduction axes are
+//! empty, and otherwise return true if any reduction axis is empty.
+bool reductionOutputShouldBeReplaced(
+    TensorView* out,
+    const std::vector<size_t>& empty_axes) {
+  bool has_empty_reduction = false;
+  bool has_empty_nonreduction = false;
+  // A reduction op itself should not generate an R-Factor domain anyway, but
```

Comment:
Please also mention that reduction rfactor should not be used at this point (if so, a reduction output can have an rfactor domain).

Should we assert that `out` doesn't have an rfactor domain?